### PR TITLE
Fix 1338

### LINKED ISSUE #1338: Funky "has not been compiled" problem
Start a fresh Factor instance, and paste this into the UI listener.

```factor
:: sieve ( n -- primes )
    n 1 + t <array> :> sieve
    f 0 sieve set-nth
    f 1 sieve set-nth

    2 n [a,b] [| i |
        i sieve nth [
            2 i * n i <range> [| j |
                f j sieve set-nth
            ] each
        ] when
    ] each

    sieve [ drop ] selector [ each-index ] dip ;
```

You will get two restarts: one for ``locals``, one for ``math.ranges``.

After choosing the "right" vocabs to use, try running ``100 sieve``.

You get this error:

```
Cannot execute a word before it has been compiled
word sieve
```

### PR DESCRIPTION
**Pull request by bjourne**
This pr should fix #1338 and probably also #1312. I just added `manifest [ clone [ clone ] change-qualified-vocabs ] change` and changes the name to `with-lambda-scope` to communicate the intent better.

### CODE REVIEW

File: basis/functors/functors.factor [hunk 1]
```diff
@@ -159,16 +153,14 @@ DEFER: ;FUNCTOR delimiter
     [
         building get use-words
         (parse-bindings)
-        building get unuse-words
     ] with-bindings ;
 
 : parse-functor-body ( -- form )
-    push-functor-words
+    functor-words use-words
```
Comment (mrjbq7) [inline basis/functors/functors.factor]: This is outside the scope of the ``with-lambda-scope`` so shouldn't we still ``unuse-words``?

File: basis/locals/parser/parser-tests.factor [hunk 2]
```diff
@@ -67,3 +55,25 @@ IN: locals.parser.tests
     ] with-compilation-unit
     [ locals>> [ name>> ] map ] [ keys ] bi*
 ] unit-test
+
+<<
```
Comment (mrjbq7) [inline basis/locals/parser/parser-tests.factor]: Does this need ``<<``?

File: basis/locals/parser/parser.factor [hunk 3]
```diff
@@ -31,15 +31,16 @@ ERROR: invalid-local-name name ;
 
 SINGLETON: lambda-parser
 
-: ((parse-lambda)) ( assoc reader-quot: ( -- quot ) -- quot )
+: with-lambda-scope ( assoc reader-quot: ( -- quot ) -- quot )
     '[
         in-lambda? on
         lambda-parser quotation-parser set
-        [ use-words @ ] [ unuse-words ] bi
+        manifest [ clone [ clone ] change-qualified-vocabs ] change
```
Comment (mrjbq7) [inline basis/locals/parser/parser.factor]: Seems like a clean fix -- another option would be to define ``M\ manifest clone`` and all four containers in it (``search-vocab-names``, ``search-vocabs``, ``qualified-vocabs``, ``auto-used``).  On the other hand, this is a simple specific fix which I like.

Comment (mrjbq7) [inline basis/locals/parser/parser.factor]: Actually, we already define ``M\ manifest clone``, so we could just change this to:

```
manifest [ clone ] change
```

**Comment by mrjbq7** [comment]
Okay, I merged this with two small fixes:

functors: fc467b72a59be1850827a3612af929dadc96c36c

locals.parser: e69d0a7b45bab455c9136912376386bfe592424e

File: basis/locals/parser/parser.factor (hunk 3 above)
Comment (bjourne) [inline basis/locals/parser/parser.factor]: Nice. I didn't see that.

File: basis/functors/functors.factor (hunk 1 above)
Comment (bjourne) [inline basis/functors/functors.factor]: I don't think so because it is only called from `parse-locals-definition` which sets up the "lambda scope" for it. I saw that you added a new scope for it and that should work too.

File: basis/locals/parser/parser-tests.factor (hunk 2 above)
Comment (bjourne) [inline basis/locals/parser/parser-tests.factor]: I'm fairly sure it's needed. `<<` and `>>` does some kind of magic so you need them when running tests that tests the parser.
